File: 01-array-string/24-text-justification/Solution.py
class Solution:
    def fullJustify(self, words: list[str], maxWidth: int) -> list[str]:
        res = []
        idx = 0
        n = len(words)
        while idx < n:
            total = len(words[idx])
            last = idx + 1
            while last < n and total + len(words[last]) + (last - idx) <= maxWidth:
                total += len(words[last])
                last += 1

            num_words = last - idx
            if last == n or num_words == 1:
                line = ' '.join(words[idx:last])
                line += ' ' * (maxWidth - len(line))
            else:
                total_spaces = maxWidth - total
                gaps = num_words - 1
                space_each, extra = divmod(total_spaces, gaps)
                parts = []
                for i in range(idx, last - 1):
                    parts.append(words[i])
                    spaces = space_each + (1 if (i - idx) < extra else 0)
                    parts.append(' ' * spaces)
                parts.append(words[last - 1])
                line = ''.join(parts)

            res.append(line)
            idx = last
        return res

File: 01-array-string/24-text-justification/Solution.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> res;
        int idx = 0, n = words.size();
        while (idx < n) {
            int total = words[idx].size(), last = idx + 1;
            while (last < n && total + (int)words[last].size() + (last - idx) <= maxWidth) {
                total += words[last].size();
                last++;
            }
            int numWords = last - idx;
            string line;
            if (last == n || numWords == 1) {
                for (int i = idx; i < last; i++) {
                    line += words[i];
                    if (i != last - 1) line += " ";
                }
                while ((int)line.size() < maxWidth) line += " ";
            } else {
                int totalSpaces = maxWidth - total;
                int gaps = numWords - 1;
                int spaceEach = totalSpaces / gaps;
                int extra = totalSpaces % gaps;
                for (int i = idx; i < last - 1; i++) {
                    line += words[i];
                    int spaces = spaceEach + ((i - idx) < extra ? 1 : 0);
                    line += string(spaces, ' ');
                }
                line += words[last - 1];
            }
            res.push_back(line);
            idx = last;
        }
        return res;
    }
};
